server/src/main/java/io/crate/user/TransportAlterRoleAction.java

```diff
@@ -112,14 +112,22 @@ static boolean alterRole(Metadata.Builder mdBuilder, String roleName, @Nullable
         if (oldRolesMetadata.roles().get(roleName).isUser() == false && secureHash != null) {
             throw new UnsupportedFeatureException("Setting a password to a ROLE is not allowed");
         }
+
+        RolesMetadata newMetadata;
         // create a new instance of the metadata, to guarantee the cluster changed action
-        // and add old UsersMetadata
-        RolesMetadata newMetadata = RolesMetadata.newInstance(oldRolesMetadata).addOldUsersMetadata(oldUsersMetadata);
+        // and user old UsersMetadata if exiting
+        if (oldUsersMetadata != null) {
+            // could be after upgrade or when users have been restored from old snapshot
+            // and we want to override all existing users & roles
+            newMetadata = new RolesMetadata().addOldUsersMetadata(oldUsersMetadata);
+            mdBuilder.removeCustom(UsersMetadata.TYPE);
+        } else {
+            newMetadata = RolesMetadata.newInstance(oldRolesMetadata);
+        }
```

Comment:
Maybe worth moving this into some static method since the code is now duplicated in all role actions?